src/components/sync-dialog.tsx

```diff
@@ -0,0 +1,214 @@
+	pull: {
+		staging: {
+			title: __( 'Pull from Staging' ),
+			description: __(
+				"Pulling will replace your Studio site's files and database with a copy from your staging site."
+			),
+		},
+		production: {
+			title: __( 'Pull from Production' ),
+			description: __(
+				"Pulling will replace your Studio site's files and database with a copy from your production site."
+			),
+		},
+		fromLabel: __( 'Pull' ),
+		toLabel: __( 'To' ),
+		subtitleSelector: __( 'What would you like to pull?' ),
+		envSync: __( 'Read more about <a>environment pull</a> ↗' ),
+		submit: __( 'Pull' ),
+	},
+	push: {
+		staging: {
+			title: __( 'Push to Staging' ),
+			description: __(
+				'Pushing will replace the existing files and database with a copy from your local site.\n\n The staging site will be backed-up before any changes are applied.'
+			),
+		},
+		production: {
+			title: __( 'Push to Production' ),
+			description: __(
+				'Pushing will replace the existing files and database with a copy from your local site.\n\n The production site will be backed-up before any changes are applied.'
+			),
+		},
+		fromLabel: __( 'Push' ),
+		toLabel: __( 'To' ),
+		subtitleSelector: __( 'What would you like to push?' ),
+		envSync: __( 'Read more about <a>environment push</a> ↗' ),
+		submit: __( 'Push' ),
+	},
+};
+
+type SyncDialogProps = {
+	type: 'push' | 'pull';
+	localSite: SiteDetails;
+	remoteSite: SyncSite;
+	onSubmit: ( syncData: TreeNode[] ) => void;
+	onRequestClose: () => void;
+};
+
+export function SyncDialog( {
+	type = 'push',
+	localSite,
+	remoteSite,
+	onSubmit,
+	onRequestClose,
+}: SyncDialogProps ) {
+	const { locale } = useI18nData();
+
+	const [ treeState, setTreeState ] = useState< TreeNode[] >( [
+		{
+			id: 'filesAndFolders',
+			label: 'Files and folders',
+			checked: true,
+			indeterminate: false,
+			defaultExpanded: false,
+			customExpanderOptions: [
+				{
+					label: __( 'All files and folders' ),
+					value: 'collapsed',
+				},
+				{
+					label: __( 'Specific files and folders' ),
+					value: 'expanded',
+				},
+			],
+			children: [
+				{
+					id: 'wp-content',
+					label: 'wp-content',
+					checked: true,
+					indeterminate: false,
+					type: 'folder',
+					children: [
+						{
+							id: 'plugins',
+							label: 'plugins',
+							checked: true,
+							type: 'folder',
+							children: [
+								{ id: 'akismet', label: 'akismet', checked: true, type: 'folder' },
+								{ id: 'jetpack', label: 'jetpack', checked: true, type: 'folder' },
+							],
+						},
+						{
+							id: 'themes',
+							label: 'themes',
+							checked: true,
+							type: 'folder',
+						},
+						{
+							id: 'uploads',
+							label: 'uploads',
+							checked: true,
+							type: 'folder',
+						},
+					],
+				},
+			],
+		},
+		{
+			id: 'database',
+			label: 'Database',
+			checked: true,
+		},
+	] );
+
+	const copy = allCopy[ type ];
+	const remoteSiteType = remoteSite.isStaging ? 'staging' : 'production';
```

Comment:
Let's take into account also the `sandbox` environment type. We can use `environmentType` to infer it.